src/mongocrypt-crypto.c

```diff
@@ -1310,4 +1325,195 @@ _mongocrypt_fle2aead_do_decryption (_mongocrypt_crypto_t *crypto,
    }
 
    return true;
-}
\ No newline at end of file
+}
+
+bool
+_mongocrypt_fle2_do_encryption (_mongocrypt_crypto_t *crypto,
+                                const _mongocrypt_buffer_t *iv,
+                                const _mongocrypt_buffer_t *key,
+                                const _mongocrypt_buffer_t *plaintext,
+                                _mongocrypt_buffer_t *ciphertext,
+                                uint32_t *bytes_written,
+                                mongocrypt_status_t *status)
+{
+   BSON_ASSERT_PARAM (crypto);
+   BSON_ASSERT_PARAM (iv);
+   BSON_ASSERT_PARAM (key);
+   BSON_ASSERT_PARAM (plaintext);
+   BSON_ASSERT_PARAM (ciphertext);
+   BSON_ASSERT_PARAM (bytes_written);
+   BSON_ASSERT_PARAM (status);
+
+   if (ciphertext->len !=
+       _mongocrypt_fle2_calculate_ciphertext_len (plaintext->len)) {
+      CLIENT_ERR ("output ciphertext must be allocated with %" PRIu32 " bytes",
+                  _mongocrypt_fle2_calculate_ciphertext_len (plaintext->len));
+      return false;
+   }
+
+   if (plaintext->len <= 0) {
+      CLIENT_ERR ("input plaintext too small. Must be more than zero bytes.");
+      return false;
+   }
+
+   if (MONGOCRYPT_IV_LEN != iv->len) {
+      CLIENT_ERR ("IV must be length %d, but is length %" PRIu32,
+                  MONGOCRYPT_IV_LEN,
+                  iv->len);
+      return false;
+   }
+   if (MONGOCRYPT_ENC_KEY_LEN != key->len) {
+      CLIENT_ERR ("key must be length %d, but is length %" PRIu32,
+                  MONGOCRYPT_ENC_KEY_LEN,
+                  key->len);
+      return false;
+   }
+
+   memset (ciphertext->data, 0, ciphertext->len);
+   *bytes_written = 0;
+
+   /* Declare variable names matching [AEAD with
+    * CTR](https://docs.google.com/document/d/1eCU7R8Kjr-mdyz6eKvhNIDVmhyYQcAaLtTfHeK7a_vE/).
+    */
+   /* M is the input plaintext. */
+   _mongocrypt_buffer_t M;
+   if (!_mongocrypt_buffer_from_subrange (&M, plaintext, 0, plaintext->len)) {
```

Comment:
This is... fine, since there's no copying involved, it just feels oddly verbose.

```
_mongocrypt_buffer_t *M = plaintext;
_mongocrypt_buffer_t *Ke = key;
_mongocrypt_buffer_t *IV = iv;
_mongocrypt_buffer_t *C = ciphertext;
```

Would achieve the same without so much bulk.

Doing the work for `S` does make sense, as it's in an offset relative to `C`, but the rest is a page with of noise that makes the whole less readable.

Same in decrypt below.

Reply:
Done.